src/main/java/io/strimzi/kafka/bridge/mqtt/mapper/MqttKafkaMapper.java

```diff
@@ -46,113 +33,61 @@ public class MqttKafkaMapper {
     public MqttKafkaMapper(List<MappingRule> rules) {
         this.rules = rules;
         this.placeholderPattern = Pattern.compile(MQTT_TOPIC_PLACEHOLDER_REGEX);
-        buildRegex();
+
+        // compile the patterns for each rule
+        this.rules.forEach(e -> this.patterns.add(Pattern.compile(e.getMqttTopicPattern())));
     }
 
     /**
      * Maps an MQTT topic to a Kafka topic. The topic is mapped according to the defined mapping rules.
      *
      * @param mqttTopic represents an MQTT topic.
-     * @return a valid Kafka topic.
+     * @return a MappingResult object containing the mapped kafka topic and kafka key.
      * @see MappingRule
+     * @see MappingResult
      */
-    public String map(String mqttTopic) {
+    public MappingResult map(String mqttTopic) {
+        MappingResult mappingResult = new MappingResult(DEFAULT_KAFKA_TOPIC, null);
```

Comment:
You are creating an instance here and modifying it later with the setter.
See my other comments, I would create the instance only at the end when you figure out which topic and key to use. It will be clearer in the code where it's created and with which values instead of having to figure out when setters are called.

Reply:
I just converted the MappingResult to a record, is that fine? We anyway do not need the setters.

Reply:
Yeah that makes sense.